content/posts/09-nibbler-flatpak.md

```diff
@@ -0,0 +1,230 @@
+by the certainty it has that the value given to the move is correct, and sorting
+the moves by the amount of moves left until a conclusion (win, draw or loss) to
+the game.
+
+Unfortunately for us it is written in Electron which makes it very hard to package
+for [musl][2] systems like [Alpine Linux][3] which I use and am a developer,
+fortunately there is [flatpak][4] which is a godsend for packaging heavyweight apps,
+so lets go package it.
+
+<!--more-->
+
+## Writing the boilerplate
+
+Lets start by writing the boilerplate we need to get started on any Flatpak app.
+
+```yaml
+app-id: org.fohristiwhirl.nibbler
+runtime: org.freedesktop.Platform
+runtime-version: '20.08'
+branch: stable
+sdk: org.freedesktop.Sdk
+command: /app/nibbler/nibbler
+
+finish-args:
+  - --share=ipc # Needed for X11
+  - --socket=x11 # Electron application, needs Ozone, switch to fallback-x11 once chrome gets ozone
+  - --device=dri # need access to the GPU
+  - --filesystem=xdg-data # Access to $XDG_DATA_HOME, letting the user load networks
+```
+
+This is very simple, and comments explain what we need every switch we need in finish-args.
+
+## Writing the module
+
+Now lets package Nibbler itself, in this case we will use the pre-packaged Nibbler
+for Glibc x86_64 instead of dealing with compiling Electron which is extremely
+annoying on Flatpak.
+
+```yaml
+modules:
+    - name: nibbler
+    buildsystem: simple
+    sources:
+      - type: archive
+        url: https://github.com/fohristiwhirl/nibbler/releases/download/v1.5.7/nibbler-1.5.7-linux.zip
+        sha256: f64d7fc89f5cd68b41dccff4fc7aa2f03a482750f2d7fa6af522f78d74264f4d
+      - type: file
+        path: org.fohristiwhirl.nibbler.desktop
+        sha256: 75ce7081b35321eeeb80043483a6977babe11055ed41f3a22cc718b63fabc6fa
+    build-commands:
+      - |
+        set -e
+        install -Dm0644 org.fohristiwhirl.nibbler.desktop -t /app/share/applications
+        mkdir -p /nibbler
+        mv * /nibbler
+        mkdir -p /app
+        mv /nibbler /app
+        chmod +x /app/nibbler/nibbler
+```
+
+### .desktop file
+
+Nibbler itself has no .desktop file, so let's write one too, one can notice that
+`--no-sandbox` is passed, that is required because Flatpak doesn't allow SUID or
+root-owned binaries, so we can't use their sandbox.
+
+```ini
+[Desktop Entry]
+Name=Nibbler (Flatpak)
+Comment=Leela Chess Zero (Lc0) Interface
+Exec=/app/nibbler/nibbler --no-sandbox %U
+Icon=Nibbler
+StartupNotify=true
+Terminal=false
+Type=Application
+Categories=Game
+```
+
+### Packaging Lc0
+
+Now the hardest part, packaging the Neural Network Engine itself, nn-engines are
+composed of two components:
+
+1. binary which runs and calculate chess positions, providing value output used
+   by our Chess GUI, and it does so with the second component.
+2. a `weights' file, which is a collection of numbers and bytes connections that
+   somehow make the binary provide good output that makes the engine very very
+   strong.
+
```

Comment:
It is hard to simplify what a neural network weights file is when you're not trained in ML.

Try simplifying this: https://www.chessprogramming.org/Leela_Chess_Zero#Network

Reply:
A bit more info on what that `weights` file actually does would be neat. I suppose it contains the "knowledge" of the engine that it obtained by being trained in ML fashion